packages/fx-core/src/component/driver/aad/update.ts

```diff
@@ -0,0 +1,142 @@
+import { Service } from "typedi";
+import { InvalidParameterUserError } from "./error/invalidParameterUserError";
+import { UpdateAadAppOutput } from "./interface/updateAadAppOutput";
+import * as fs from "fs-extra";
+import * as path from "path";
+import { AadAppClient } from "./utility/aadAppClient";
+import axios from "axios";
+import { SystemError, UserError, ok, err, FxError, Result } from "@microsoft/teamsfx-api";
+import { UnhandledSystemError, UnhandledUserError } from "./error/unhandledError";
+import { getUuid } from "../../../common/tools";
+import { expandEnvironmentVariable } from "../../utils/common";
+import { AadManifestHelper } from "../../resource/aadApp/utils/aadManifestHelper";
+import { AADManifest } from "../../resource/aadApp/interfaces/AADManifest";
+import { MissingFieldInManifestUserError } from "./error/invalidFieldInManifestError";
+import isUUID from "validator/lib/isUUID";
+
+const actionName = "aadApp/update"; // DO NOT MODIFY the name
+const helpLink = "https://aka.ms/teamsfx-actions/aadapp-update";
+
+// logic from src\component\resource\aadApp\aadAppManifestManager.ts
+@Service(actionName) // DO NOT MODIFY the service name
+export class UpdateAadAppDriver implements StepDriver {
+  public async run(
+    args: UpdateAadAppArgs,
+    context: DriverContext
+  ): Promise<Result<Map<string, string>, FxError>> {
+    try {
+      this.validateArgs(args);
+      const aadAppClient = new AadAppClient(context.m365TokenProvider);
+      const state = this.loadCurrentState();
+
+      const manifest = await this.loadManifest(args.manifestPath, state);
+      const warningMessage = AadManifestHelper.validateManifest(manifest);
+      if (warningMessage) {
+        warningMessage.split("\n").forEach((warning) => {
+          context.logProvider?.warning(warning);
+        });
+      }
+
+      if (!manifest.id || !isUUID(manifest.id)) {
+        throw new MissingFieldInManifestUserError(actionName, "id", helpLink);
+      }
+
+      // MS Graph API does not allow adding new OAuth permissions and pre authorize it within one request
+      // So split update AAD app to two requests:
+      // 1. If there's preAuthorizedApplications, remove it temporary and update AAD app to create possible new permission
+      if (manifest.preAuthorizedApplications && manifest.preAuthorizedApplications.length > 0) {
+        const preAuthorizedApplications = manifest.preAuthorizedApplications;
+        manifest.preAuthorizedApplications = [];
+        await aadAppClient.updateAadApp(manifest);
+        manifest.preAuthorizedApplications = preAuthorizedApplications;
+      }
+      // 2. Update AAD app again with full manifest to set preAuthorizedApplications
+      await aadAppClient.updateAadApp(manifest);
+
+      // Output actual manifest to project folder
+      await fs.ensureDir(path.dirname(args.manifestOutputPath));
+      await fs.writeFile(args.manifestOutputPath, JSON.stringify(manifest, null, 4), "utf8");
```

Comment:
manifestOutputPath actually is a file, so maybe manifestOutputFile is easy to understand?

Reply:
updated to outputFilePath after synced with @jidddddd